Fix SelectControl focus and de-focus bug

ilyasfoo opened the pull request:
Fixes #6879

- Fixes `<SelectControl>` component from unintended selecting the first option when focusing and de-focusing (such as navigating forms via Tab/Shift-Tab).
- Also changes the behaviour to automatically scroll to the currently selected option when the options list is displayed.

### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

-   [x] I've tested using only a keyboard (no mouse)
-   [ ] I've tested using a screen reader

### Screenshots

Automatic scroll:

#### Before:
<img src="https://user-images.githubusercontent.com/3747241/116519746-6edef180-a904-11eb-88ba-c7237e19bcae.png" width="400">

#### After:
<img src="https://user-images.githubusercontent.com/3747241/116520508-676c1800-a905-11eb-9159-a6e6eeeede1c.png" width="400">


### Detailed test instructions:

1. Navigate to Setup Wizard.
2. Select any country in `Country / Region` except for the first option (Afghanistan).
3. Set focus to `Address line 2` by clicking on the text field.
4. Press `Tab` twice, so that the Country field is focused and de-focused.
5. Observe the selected option did not change to Afghanistan.

ilyasfoo committed:
Fix select-control component to preserve currently selected item on focus and de-focus

ilyasfoo committed:
Add test

ilyasfoo committed:
Added changelog

jeffstieler commented on packages/components/src/select-control/index.js:
```diff
@@ -25,10 +25,15 @@ const initialState = { isExpanded: false, isFocused: false, query: '' };
+			selectedIndex:
+				selected && options?.length
+					? options.findIndex( ( option ) => option.key === selected )
+					: 0,
```
What do you think about defaulting to no selection (`-1`)?

That way, when a user tabs over the control without interacting, nothing is selected unexpectedly.

ilyasfoo replied:
Yep makes sense, changes made in 168c4c4.

jeffstieler reviewed (changes requested):
Overall this is a great change and does solve #6879, but I think there is an opportunity to refine the UX just a bit.

ilyasfoo committed:
No longer auto select option on initial render, pressing tab would only select if selection exists

ilyasfoo commented on packages/components/src/select-control/index.js:
```diff
@@ -233,7 +249,7 @@ export class SelectControl extends Component {
 				query,
 				isFocused: true,
-				selectedIndex: 0,
+				selectedIndex: query?.length > 0 ? 0 : this.state.selectedIndex, // Only set to 0 if we're actually searching.
```
That behaviour makes more sense, changes made 168c4c4. Thanks for making the suggestion!

ilyasfoo committed:
Merge remote-tracking branch 'origin/main' into fix/6879-select-control-selected-bugs

jeffstieler commented on packages/components/src/select-control/index.js:
```diff
@@ -25,15 +25,10 @@ const initialState = { isExpanded: false, isFocused: false, query: '' };
-				selected && options?.length
-					? options.findIndex( ( option ) => option.key === selected )
-					: 0,
+			selectedIndex: -1,
```
This seems to break `SelectControl`s that are supplied a `selected` prop. I think all we need to change from the original code is the `selectedIndex` from `0` to `-1`.

ilyasfoo replied:
Sorry Jeff I don't get this one, this change _is from `0` to `-1`_ from the original code. I've also tested a `SelectControl` with `selected` prop given, and seems working fine with mine. If it's possible, maybe you can help to provide an example where it breaks? :pray:

I'll make an update to the storybook to include an example with `selected` as well.

ilyasfoo replied:
Ah, I misunderstood your original suggestion 🤦 . Changes made in 0890a7b

jeffstieler commented on packages/components/src/select-control/list.js:
```diff
@@ -116,9 +116,11 @@ class List extends Component {
-				this.select( options[ selectedIndex ] );
-				setExpanded( false );
-				onSearch( null );
+				if ( selectedIndex > -1 ) {
```
Given the other changes, I don't think this check is needed.

ilyasfoo replied:
Actually we do, since pressing tab on the component will attempt to select anyways. Doesn't break the component but it does prompt an error on [this line](https://github.com/woocommerce/woocommerce-admin/blob/d167204498f57a74b45ae1a668050ed8674d6bb8/packages/components/src/select-control/list.js#L48) since the `option` is `undefined`.

jeffstieler commented on packages/components/src/select-control/test/index.js:
```diff
@@ -260,33 +260,27 @@ describe( 'SelectControl', () => {
-				// selected option should be sufficient to simulate the logic in this test.
-				userEvent.click( getByRole( 'option', { selected: true } ) );
-				expect( onChangeMock ).toHaveBeenCalledTimes( 0 );
-			} );
```
This test case should still pass, but broke because of the changes to the initial state above.

ilyasfoo replied:
True, this is now fixed in 0890a7b

jeffstieler commented on packages/components/src/select-control/test/index.js:
```diff
@@ -260,33 +260,27 @@ describe( 'SelectControl', () => {
-				userEvent.click( getByRole( 'option', { selected: true } ) );
-				expect( onChangeMock ).toHaveBeenCalledTimes( 0 );
-			} );
+		it( 'should not have a selected option on initial render', async () => {
```
The control _should_ have a selected option on initial render if supplied a `selected` prop that matches an option's `key` should it not?

ilyasfoo replied:
This test case was the result of my misunderstanding 😅 . I've changed it to `should not automatically select first option on focus`

jeffstieler reviewed (changes requested):
Better than before, but I think we introduced some regressions with the initial state changes.

It would probably be beneficial to add a `SelectControl` to the storybook that has a value for the `selected` prop to make sure things are working as expected for that case as well.

ilyasfoo committed:
Fix to reflect the proper suggested changes

ilyasfoo commented:
Thanks for the review @jeffstieler! I've misunderstood your original suggestion, now it should work like what you'd expect. Also made updated the storybook with an example of `selected` prop. Please have another look 🙏 !

jeffstieler reviewed (approved):
This is now working as expected, thanks @ilyasfoo!

Not a blocker for this PR, but there seems to be a more noticeable delay when tabbing out of the control now. I think we're probably doing (and then undoing) some unnecessary work in our `reset()` functionality.

ilyasfoo committed:
Attempt to fix unnecessary search reset on pressing tab

ilyasfoo committed:
Fix for keyboard interactions and added accompanying tests

ilyasfoo committed:
Merge remote-tracking branch 'origin/main' into fix/6879-select-control-selected-bugs

ilyasfoo commented:
@jeffstieler Thanks for highlighting that! I've decided to look into it and found out that the [`onSearch( null )`](https://github.com/woocommerce/woocommerce-admin/pull/3507/files#diff-1e69220e8f190ce9004b709e77983470ebf919057d810c4c27b0515b26061069R122)  call here might not be necessary which caused the appearance of delay. I've removed it and did not notice any regression from the original PR, nor found anything amiss with tab behaviour in all SelectControl I could find.

On my way towards that, I've also discovered multiple regressions that have been caused by this PR:
1. When we've already selected something, search, then press up or down, it will use the index based on all options instead of the filtered ones.
2. Pressing enter after search triggers an error if nothing is selected.
3. Option list was not hidden on pressing tab without selecting anything even after losing focus on combobox (happened in compare products search).

I've made a PR to fix these and added some tests to go along with it. Hopefully, these tests should capture some of the behaviours that could easily break without us noticing. 

Would appreciate if you can have a look at this one last time!

becdetat commented:
The speed/delay of tabbing through the control seems good to me atm

becdetat reviewed (approved):
Looks good to me 👍
